cdap-app-fabric/src/main/java/co/cask/cdap/internal/profile/AdminEventPublisher.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.profile;
+
+import co.cask.cdap.api.app.ApplicationSpecification;
+import co.cask.cdap.api.messaging.MessagingContext;
+import co.cask.cdap.api.messaging.TopicNotFoundException;
+import co.cask.cdap.api.retry.RetryableException;
+import co.cask.cdap.common.ServiceUnavailableException;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.service.Retries;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.common.service.RetryStrategy;
+import co.cask.cdap.data2.metadata.writer.MetadataMessage;
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProfileId;
+import co.cask.cdap.proto.id.ScheduleId;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.common.base.Throwables;
+import com.google.gson.Gson;
+import com.google.gson.JsonNull;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class for publishing the profile metadata change request to tms
+ */
+public class AdminEventPublisher {
+
+  private static final Gson GSON = new Gson();
+
+  private final TopicId topic;
+  private final RetryStrategy retryStrategy;
+
+  @Inject
+  AdminEventPublisher(CConfiguration cConf) {
+    this.topic = NamespaceId.SYSTEM.topic(cConf.get(Constants.Metadata.MESSAGING_TOPIC));
+    this.retryStrategy = RetryStrategies.timeLimit(
+      20, TimeUnit.SECONDS,
+      RetryStrategies.exponentialDelay(10, 200, TimeUnit.MILLISECONDS));
+  }
+
+  /**
+   * Update the profile metadata for the affected programs/schedules, the metadata of programs/schedules will be
+   * updated according to the closest level preference that contains the profile information.
+   *
+   * @param messagingContext message context to get the publisher to publish the message
+   * @param entityId the entity id to update metadata, all programs/schedules under this entity id will be affected
+   * @param profileId the profileId which get assigned
+   */
+  public void publishProfileAssignment(MessagingContext messagingContext, EntityId entityId, ProfileId profileId) {
```

Comment:
Is there a reason each method needs to take a MessagingContext instead of injecting it into the constructor?

Reply:
I changed the constructor so we removed the MessagingContext as a parameter